### PR TITLE
HSL DateTime: timestamps

### PR DESCRIPTION
**Pull request by jjergus**
Adds `DateTime\Timestamp` and `MonoTime\Timestamp` + any base classes/traits they depend on.

I'm still working on tests for everything, I'll probably put them in a separate PR when ready since this is already quite long.

### CODE REVIEW

File: src/datetime/_Private/Comparable.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    return $this->compare($other) >= 0;
+  }
+
+  /**
+   * Returns true if this instance represents a point in time after `$a` but
+   * before `$b`, or vice-versa (before `$a` but after `$b`), or if this
+   * instance is equal to `$a` and/or `$b`. Returns false if this instance is
+   * before or after both `$a` and `$b`.
+   */
+  final public function isBetweenIncl(
```
Comment (azjezz) [inline src/datetime/_Private/Comparable.php:73]:
```suggestion
  final public function isBetweenInclusive(
```

more descriptive :)

File: src/datetime/_Private/Comparable.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    return $a !== 0 && $b !== 0 && $a !== $b;
+  }
+
+  /**
+   * Returns true if this instance represents a point in time after `$a` but
+   * before `$b`, or vice-versa (before `$a` but after `$b`). Returns false if
+   * this instance is equal to `$a` and/or `$b`, or if this instance is before
+   * or after both `$a` and `$b`.
+   */
+  final public function isBetweenExcl(
```
Comment (azjezz) [inline src/datetime/_Private/Comparable.php:88]:
```suggestion
  final public function isBetweenExclusive(
```

same

File: src/datetime/FormatString.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace HH\Lib\Experimental\DateTime {
+  /**
+   * strftime(3)-compatible format string that includes standard placeholders
+   * like `%y` for year, but does not include timezone placeholders like `%z`.
+   */
+  type UnzonedDateFormatString = \HH\FormatString<_Private\DateFormat>;
```
Comment (fredemmott) [inline src/datetime/FormatString.php:16]: This looks wrong - shouldn't this be \HH\Lib\_Private\_DateTime\DateFormat?

File: src/datetime/FormatString.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * strftime(3)-compatible format string that includes standard placeholders
+   * like `%y` for year, as well as timezone placeholders like `%z`.
+   */
+  type ZonedDateFormatString = \HH\FormatString<_Private\ZonedDateFormat>;
+}
+
+namespace HH\Lib\Experimental\DateTime\_Private {
```
Comment (fredemmott) [inline src/datetime/FormatString.php:25]: ah yep :)

Also, split out to separate file, at least one per NS if not per class.

- generally good to separate private from non-private where possible
- we currently have tools (e.g. the flib wrapper generator) that messes up if there are multiple NSes in a file

**Review by fredemmott — APPROVED** [review]

**Pull request merged**